### PR TITLE
Address known limitation in service/client

### LINKED ISSUE #105: Address known service/client (request/ response) limitation
Circumvent the [known limitation](https://github.com/ros2/rmw_iceoryx/releases/tag/v2.0.3-iron) by storing the sequence together with the sample pointer in a `std::map` as discussed [here](https://github.com/ros2/rmw_iceoryx/pull/98#discussion_r1328709763).

### PR DESCRIPTION
**Pull request by mossmaurice**
Closes #105

### CODE REVIEW

File: rmw_iceoryx_cpp/src/rmw_response.cpp [hunk 1]
```diff
@@ -147,14 +147,14 @@ rmw_send_response(
...
-  if (!iceoryx_server_abstraction->request_payload_) {
+  if (iceoryx_server_abstraction->request_payload_map_.empty()) {
     RMW_SET_ERROR_MSG("'rmw_take_request' needs to be called before 'rmw_send_response'!");
     ret = RMW_RET_ERROR;
     return ret;
   }
 
   auto * iceoryx_request_header = iox::popo::RequestHeader::fromPayload(
-    iceoryx_server_abstraction->request_payload_);
+    iceoryx_server_abstraction->request_payload_map_[request_header->sequence_number]);
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply rmw_iceoryx_cpp/src/rmw_response.cpp]: @elBoberido Thanks for posting the article, good point. Fixed.

File: rmw_iceoryx_cpp/src/rmw_response.cpp [hunk 2]
```diff
@@ -186,9 +186,11 @@ rmw_send_response(
       ret = RMW_RET_ERROR;
     });
 
-  // Release the hold request
-  iceoryx_server->releaseRequest(iceoryx_server_abstraction->request_payload_);
-  iceoryx_server_abstraction->request_payload_ = nullptr;
+  // Release the hold request and delete the entry from the map
+  iceoryx_server->releaseRequest(
+    iceoryx_server_abstraction->request_payload_map_[request_header->
+    sequence_number]);
```
[the remark this answers is not in the document]
Comment (mossmaurice) [reply rmw_iceoryx_cpp/src/rmw_response.cpp]: @elBoberido I think using the `RequestHeader*` to release the chunk is not what `UntypedServerImpl<BaseServerT>::releaseRequest` is expecting. We have to use the original payload pointer, right?

**Review by elBoberido — APPROVED** [review]

File: rmw_iceoryx_cpp/src/rmw_response.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (elBoberido) [reply rmw_iceoryx_cpp/src/rmw_response.cpp]: Yes, you are right. I overlooked the translation of the pointer.

**Comment by elBoberido** [comment]
Looks good. LGTM

**Pull request merged**